### PR TITLE
Adding Multi-Line, RTL Text, and Better Caret Support to Keyboard Preview and Non-Native Keyboard

### PR DESCRIPTION
**Pull request by AMollis**
## Overview
This adds better multi-line and caret support to the keyboard preview in MRTK3.  Also, the non-native keyboard had no caret/cursor on HL2.  This change adds the caret to the non-native keyboard on HL2.

keyboard-preview

## Changes
- Fixes: https://github.com/microsoft/MixedRealityToolkit-Unity/issues/11558

## Verification
- Tested on HL2 with native a non-native keyboard
- Created additional unit test

### CODE REVIEW

File: com.microsoft.mrtk.uxcore/Keyboard/KeyboardPreview.cs [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+
+
+        private void UpdateLayout()
+        {
+            if (previewCaret == null || PreviewText == null)
+            {
+                return;
+            }
+
+            var textInfo = PreviewText.GetTextInfo(Text);
```
Comment (shaynie) [inline com.microsoft.mrtk.uxcore/Keyboard/KeyboardPreview.cs:166]: Per frame allocation.  Can be a field var?

File: com.microsoft.mrtk.uxcore/Keyboard/KeyboardPreview.cs [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+            }
+
+            if (prevChar.lineNumber > textInfo.lineCount)
+            {
+                ResetCaret();
+                ResetView();
+                return;
+            }
+
+            var lineInfo = textInfo.lineInfo[prevChar.lineNumber];
```
Comment (shaynie) [inline com.microsoft.mrtk.uxcore/Keyboard/KeyboardPreview.cs:217]: Per frame allocation.  Can be a field var?

**Review by shaynie — CHANGES_REQUESTED** [review]

File: com.microsoft.mrtk.uxcore/Keyboard/KeyboardPreview.cs (hunk 1 above)
Comment (AMollis) [reply com.microsoft.mrtk.uxcore/Keyboard/KeyboardPreview.cs:166]: TextInfo is a struct.  no heap allocation.  

This is needed unfortunately. TextInfo needs to be filled in here to start layout.  UpdatLayout() only happen with new keyboard input

File: com.microsoft.mrtk.uxcore/Keyboard/KeyboardPreview.cs (hunk 2 above)
Comment (AMollis) [reply com.microsoft.mrtk.uxcore/Keyboard/KeyboardPreview.cs:216]: no allocation here

**Review by shaynie — APPROVED** [review]

**Pull request merged**